src/coreclr/src/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -199,12 +199,112 @@ internal static bool HasElementType(RuntimeType type)
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern object CreateInstance(RuntimeType type, bool publicOnly, bool wrapExceptions, ref bool canBeCached, ref RuntimeMethodHandleInternal ctor, ref bool hasNoDefaultCtor);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        internal static extern object Allocate(RuntimeType type);
-
         [MethodImpl(MethodImplOptions.InternalCall)]
         internal static extern object CreateInstanceForAnotherGenericParameter(RuntimeType type, RuntimeType genericParameter);
 
+        /// <summary>
+        /// Given a RuntimeType, returns both the address of the JIT's newobj allocator helper for
+        /// that type and the MethodTable* corresponding to that type. Return value signature is
+        /// managed calli (MethodTable* pMT) -> object.
+        /// </summary>
+        internal static delegate*<MethodTable*, object> GetAllocatorFtn(
+            // This API doesn't call any constructors, but the type needs to be seen as constructed.
+            // A type is seen as constructed if a constructor is kept.
+            // This obviously won't cover a type with no constructor. Reference types with no
+            // constructor are an academic problem. Valuetypes with no constructors are a problem,
+            // but IL Linker currently treats them as always implicitly boxed.
+            [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors | DynamicallyAccessedMemberTypes.NonPublicConstructors)] RuntimeType type,
+            out MethodTable* pMT, bool forGetUninitializedObject, bool wrapExceptions)
+        {
+            Debug.Assert(type != null);
+
+            delegate*<MethodTable*, object> pNewobjHelperTemp = null;
+            MethodTable* pMTTemp = null;
+            Interop.BOOL fFailedWhileRunningCctor = Interop.BOOL.FALSE;
+
+            try
+            {
+                GetAllocatorFtn(
+                    new QCallTypeHandle(ref type),
+                    &pNewobjHelperTemp,
+                    &pMTTemp,
+                    forGetUninitializedObject ? Interop.BOOL.TRUE : Interop.BOOL.FALSE,
+                    &fFailedWhileRunningCctor);
+            }
+            catch (Exception ex)
+            {
+                // If the cctor failed, propagate the exception as-is, wrapping in a TIE
+                // if needed. Otherwise, make the error message friendlier by including
+                // the name of the type that couldn't be instantiated.
+
+                if (fFailedWhileRunningCctor != Interop.BOOL.FALSE)
```

Comment:
To make this easier to follow, it may be a good idea to move this outside this try/catch.

```
try
{
    RuntimeHelpers._RunClassConstructor(type, preciseOnly: true); // New argument for `_RunClassConstructor`
}
 catch (Exception e) when (wrapExceptions)
{
   throw new TargetInvocationException(ex);
}
```

Reply:
Noting the existing behavior here for reference so that there's compat documentation.

- `GetUninitializedObject` always runs precise cctors explicitly, no TIE.
- `CreateInstance` runs precise cctors explicitly _only if_ it's a value type with no parameterless ctor, no TIE.
- For types with a parameterless ctor, `CreateInstance` doesn't run the precise cctor explicitly; it instead ignores the cctor and assumes that the ctor will invoke the cctor as needed. TIE behavior is controlled by `BindingFlags.DoNotWrapExceptions`.

Reply:
> CreateInstance runs precise cctors explicitly only if it's a value type with no user-defined parameterless ctor,

This sounds like an outlier. Constructing a value type with no user-defined parameterless ctor does not trigger cctors anywhere else in the runtime. E.g.: this won't print `CCtor called!`:

```
using System;

Console.WriteLine(new StructWithCCtor().ToString());

public struct StructWithCCtor
{
    static StructWithCCtor() { Console.WriteLine("CCtor called!"); }
}
```

Reply:
Also, this is documented in ECMA spec: "Thus for value types, the .cctor is only guaranteed to be called for instances of the value type that are not “all-zero”."

Reply:
Yeah, just re-read §I.8.9.5 and I agree with your assessment. Will simplify the logic and remove the outlier.